CalPatRec/src/PhiClusterFinder_module.cc

```diff
@@ -0,0 +1,566 @@
+    void fillTimeSpectrumAlt(const std::vector<StrawHitIndex>& ordchcol);
+    void findPeaks  (TimeCluster& tc, const std::vector<StrawHitIndex>& ordchcol);
+    void assignHits(TimeCluster& otc, const std::vector<StrawHitIndex>& ordchcol);
+    void initCluster(TimeCluster& tc);
+    float checkdelta(TimeCluster& tc, int ClustNo, const std::vector<StrawHitIndex>& ordchcol);
+    void addCaloClusters(TimeClusterCollection& tccolnew, const TimeClusterCollection& tccol);
+};
+
+ PhiClusterFinder::PhiClusterFinder(const art::EDProducer::Table<Config>& config):
+    art::EDProducer{config},
+    _diag         (config().diag()),
+    _debug        (config().debug()),
+    _minnsh       (config().minNSH()),
+    _threshold    (config().threshold()),
+    _nphiclusters (config().nPhiClusters()),
+    _phimin       (config().phiMin()),
+    _phimax       (config().phiMax()),
+    _nphibins     (config().nPhiBins()),
+    _mindphi      (config().minDPhi()),
+    _pitch        (config().pitch()),
+    _tmin         (config().tMin()),
+    _tmax         (config().tMax()),
+    _tbin         (config().tBin()),
+    _t0calc       (config().t0Calc()),
+    _ymin         (config().yMin()),
+    _maxdt        (config().maxDt()),
+    _minsigma     (config().minSigma()),
+    _npeak        (config().peakWidth()),
+    _chToken      {consumes<ComboHitCollection>(config().ComboHitCollection()) },
+    _tcToken      {consumes<TimeClusterCollection>(config().TimeClusterCollection()) },
+    _useCC        (config().useCC())
+    {
+      art::ServiceHandle<art::TFileService> tfs;
+      _hist1 = tfs->make<TH1F>( "hist1" , "phi spectrum",_nphibins,_phimin,_phimax);
+      unsigned nbinst = (unsigned)rint((_tmax-_tmin)/_tbin);
+      _timespec = TH1F("timespec","time spectrum",nbinst,_tmin,_tmax);
+      produces<TimeClusterCollection>();
+
+      if (_diag != 0) _hmanager = art::make_tool<ModuleHistToolBase>(config().DiagPlugin," ");
+      else            _hmanager = std::make_unique<ModuleHistToolBase>();
+    }
+
+  void PhiClusterFinder::beginJob() {
+    if (_diag > 0) {
+      art::ServiceHandle<art::TFileService> tfs;
+      _hmanager->bookHistograms(tfs);
+    }
+  }
+
+  void PhiClusterFinder::produce(art::Event & event ){
+    _iev = event.id().event();
+    std::unique_ptr<TimeClusterCollection> tccol1(new TimeClusterCollection);
+    auto const& chH = event.getValidHandle(_chToken);
+    _chcol = chH.product();
+
+    // Save the sum phi of a cluster to calculate delta phi between the clusters
+    _phitotal.clear();
+    _nhitotal.clear();
+    _phitotal.resize(_nphiclusters);
+    _nhitotal.resize(_nphiclusters);
+
+    for(int m=0;m<_nphiclusters;m++){
+      _phitotal[m] = 0.;
+      _nhitotal[m] = 0.;
+    }
+    auto const& tcH = event.getValidHandle(_tcToken);
+    const TimeClusterCollection& tccol(*tcH);
+    // For the diagnostics
+    _data._tccol =const_cast<TimeClusterCollection*>(tcH.product());
```

Comment:
As I wrote in the other comment, `._tccol` should be a `const TimeClusterCollection*` to avoid this `const_cast`

Reply:
Yes, this has been changed. [const TimeClusterCollection*](https://github.com/NamithaChitrazee/Offline/commit/6da4c6b6b23a5d9cdae1c3d40d8475aea48a731b#diff-fdcfc4df7cada41e0f990c317d0561385316cce5ccc0ab47414f76a89d1f29b3:~:text=_data._tccol%20%3D%20tcH.product()%3B)